src/metaswitch/clearwater/etcd_shared/plugin_utils.py

```diff
@@ -24,36 +24,49 @@ def run_command(command_args, namespace=None, log_error=True):
     in as an array instead of a string.
     """
-    if namespace:
-        command_args[0:0] = ['ip', 'netns', 'exec', namespace]
-
-    # Pass the close_fds argument to avoid the pidfile lock being held by
-    # child processes
-    p = subprocess.Popen(command_args,
-                         stdout=subprocess.PIPE,
-                         stderr=subprocess.PIPE,
-                         close_fds=True)
-    stdout, stderr = p.communicate()
-    if p.returncode != 0:
-        # it failed, log the return code and output
-        if log_error:
-            _log.error("Command {} failed with return code {}, "
-                       "stdout {!r}, and stderr {!r}".format(' '.join(command_args),
-                                                             p.returncode,
-                                                             stdout,
-                                                             stderr))
-            return p.returncode
-    else:
-        # it succeeded, log out stderr of the command run if present
-        if stderr:
-            _log.warning("Command {} succeeded, with stderr output {!r}".
-                         format(' '.join(command_args), stderr))
+    processes = []
+    error_returncodes = []
+    for command_args in list_of_command_args:
+        if namespace:
+            command_args[0:0] = ['ip', 'netns', 'exec', namespace]
+
+        # Pass the close_fds argument to avoid the pidfile lock being held by
+        # child processes
+        p = subprocess.Popen(command_args,
+                             stdout=subprocess.PIPE,
+                             stderr=subprocess.PIPE,
+                             close_fds=True)
+        processes.append(p)
+
+    for p in processes:
+        stdout, stderr = p.communicate()
+        if p.returncode != 0:
+            # it failed, log the return code and output
+            if log_error:
+                _log.error("Command {} failed with return code {}, "
+                           "stdout {!r}, and stderr {!r}".format(' '.join(command_args),
+                                                                 p.returncode,
+                                                                 stdout,
+                                                                 stderr))
+            error_returncodes.append(p.returncode)
         else:
-            _log.debug("Command {} succeeded".format(' '.join(command_args)))
+            # it succeeded, log out stderr of the command run if present
+            if stderr:
+                _log.warning("Command {} succeeded, with stderr output {!r}".
+                             format(' '.join(command_args), stderr))
+            else:
+                _log.debug("Command {} succeeded".format(' '.join(command_args)))
 
+    if error_returncodes:
+        return error_returncodes[0]
+    else:
         return 0
 
 
+def run_command(command_args, namespace=None, log_error=True):
```

Comment:
No docstring. You're using an interesting design pattern here (the function to run a single command being a call to the one that runs multiple), so it's worth explaining why you've chosen to use it.

Reply:
Is it actually that interesting or worth commenting on?  It feels like the obvious thing to me - it's just the special case of `run_commands` where N=1, and anything else would mean duplicated code.

Reply:
I'd have thought the obvious way round to do something like this is:
- `run_command()` contains the code for running a single command
- `run_commands()` contains the code for iterating `run_command()` as necessary

There are good reasons why you chose not to do it that way, but if I were coming to the code fresh I might expect to find how a command is being run under `run_command` and that's not obvious here.